### PR TITLE
feat: Add ConfigCat provider

### PR DESCRIPTION
**Pull request by luizbon**
## This PR
- Adding ConfigCat provider integration

### How to test
Have a ConfigCat free account, create Product/Config/Keys and use as any other OpenFeature provider.

### CODE REVIEW

File: src/OpenFeature.Contrib.ConfigCat/ConfigCatProvider.cs
Comment (laliconfigcat) [inline src/OpenFeature.Contrib.ConfigCat/ConfigCatProvider.cs:80]: As I see this check is unnecessary as the next .Contains() will do the same.

File: src/OpenFeature.Contrib.ConfigCat/ConfigCatProvider.cs
Comment (laliconfigcat) [inline src/OpenFeature.Contrib.ConfigCat/ConfigCatProvider.cs:78]: As the error messages can change much easier, I'd go with checking for the error codes and levels:
https://github.com/configcat/.net-sdk/blob/master/src/ConfigCatClient/Logging/LogMessages.cs#L10
The format looks like this: https://github.com/configcat/.net-sdk/blob/master/src/ConfigCat.Client.Tests/data/evaluationlog/simple_value/int_setting.txt#L1

So e.g. instead of "Config JSON is not present", I'd check for "ERROR [1000]"

File: src/OpenFeature.Contrib.ConfigCat/ConfigCatProvider.cs
Comment (laliconfigcat) [inline src/OpenFeature.Contrib.ConfigCat/ConfigCatProvider.cs:73]: As the GetValueDetailsAsync handling everything  - it returns the defaultValue if needed, I'd not override this behaviour in any cases with giving back the defaultValue if the errorMessage is not empty.

I think this return can be simplified to 
`return new ResolutionDetails<T>(flagKey, result.Value, !string.IsNullOrEmpty(result.ErrorMessage) ? ParseErrorType(result.ErrorMessage) : null, variant: result.VariationId, errorMessage: result.ErrorMessage);`

File: src/OpenFeature.Contrib.ConfigCat/UserBuilder.cs
Comment (laliconfigcat) [inline src/OpenFeature.Contrib.ConfigCat/UserBuilder.cs:18]: I'd only initialize the new User with a GUID if you could not parse the "id" or "identifier" from the context.

**Review by laliconfigcat — CHANGES_REQUESTED** [review]
Hello @luizbon ,

Thank you very much for your contribution! We really appreciate it at ConfigCat.
I left some minor comments, but I think it looks good.

Have a nice day!

**Comment by toddbaert** [comment]
I'll give this a review tomorrow! Thanks @luizbon

File: src/OpenFeature.Contrib.ConfigCat/ConfigCatProvider.cs
Comment (luizbon) [reply src/OpenFeature.Contrib.ConfigCat/ConfigCatProvider.cs:80]: oops... 🤣

File: src/OpenFeature.Contrib.ConfigCat/ConfigCatProvider.cs
Comment (luizbon) [reply src/OpenFeature.Contrib.ConfigCat/ConfigCatProvider.cs:78]: Yeah, I would love to use the error codes, but unfortunately the `ErrorMessage` does not contain it.
I also checked the exception object, and it does not contain the error code.
This code is only used in the Logger.
image

File: src/OpenFeature.Contrib.ConfigCat/ConfigCatProvider.cs
Comment (luizbon) [reply src/OpenFeature.Contrib.ConfigCat/ConfigCatProvider.cs:73]: Good suggestion.

File: src/OpenFeature.Contrib.ConfigCat/UserBuilder.cs
Comment (luizbon) [reply src/OpenFeature.Contrib.ConfigCat/UserBuilder.cs:18]: Fair enough

File: src/OpenFeature.Contrib.ConfigCat/ConfigCatProvider.cs
Comment (laliconfigcat) [reply src/OpenFeature.Contrib.ConfigCat/ConfigCatProvider.cs:76]: OMG, I'm really sorry, I totally thought that we are giving back the error code too. 
I'll talk about this with the guys. Totally makes sense to give back the error code too.

**Comment by luizbon** [comment]
Hi @laliconfigcat, I have addressed the points you mentioned. Thanks for the feedback.
There's only one pending regarding the error code, though.

**Review by laliconfigcat — APPROVED** [review]
Really good job! Thank you again for your efforts and for the quick changes.

File: src/OpenFeature.Contrib.ConfigCat/ConfigCatProvider.cs
Comment (toddbaert) [reply src/OpenFeature.Contrib.ConfigCat/ConfigCatProvider.cs:76]: I guess a new issue can be created if this is added to the config cat SDK.

File: test/OpenFeature.Contrib.ConfigCat.Test/ConfigCatProviderTest.cs
Comment (toddbaert) [inline test/OpenFeature.Contrib.ConfigCat.Test/ConfigCatProviderTest.cs:98]: This definitely makes testing this provider easy!

File: src/OpenFeature.Contrib.ConfigCat/ConfigCatProvider.cs
Comment (toddbaert) [inline src/OpenFeature.Contrib.ConfigCat/ConfigCatProvider.cs:69]:
```suggestion
        private async Task<ResolutionDetails<T>> ResolveFlag<T>(string flagKey, EvaluationContext context, T defaultValue)

```

Nit: "resolve" is the term OpenFeature likes to use, so I'd recommend that name (though since this is a private function, it's really up to you.

File: src/OpenFeature.Contrib.ConfigCat/ConfigCatProvider.cs
Comment (toddbaert) [inline src/OpenFeature.Contrib.ConfigCat/ConfigCatProvider.cs:73]: With this SDK (and most others) the way to indicate an error is to throw an [exception](https://github.com/open-feature/dotnet-sdk-contrib/tree/main/src/OpenFeature.Contrib.Providers.GOFeatureFlag/exception). This is particularly important because [error hooks](https://openfeature.dev/specification/sections/hooks) only run if an error is thrown. I would amend this logic so that if an error is detected, you throw instead.

I understand this is a bit confusing. We should probably make it impossible to directly _return_ an error code in from the Resolvers.

File: src/OpenFeature.Contrib.ConfigCat/UserBuilder.cs
Comment (toddbaert) [inline src/OpenFeature.Contrib.ConfigCat/UserBuilder.cs:13]: It seems this isn't covered by tests.

File: src/OpenFeature.Contrib.ConfigCat/UserBuilder.cs
Comment (toddbaert) [inline src/OpenFeature.Contrib.ConfigCat/UserBuilder.cs:11]: It may be worth documenting how this mapping works in a README.

File: src/OpenFeature.Contrib.ConfigCat/UserBuilder.cs
Comment (toddbaert) [inline src/OpenFeature.Contrib.ConfigCat/UserBuilder.cs:37]: Same as https://github.com/open-feature/dotnet-sdk-contrib/pull/119/files#r1438292902

**Review by toddbaert — CHANGES_REQUESTED** [review]
Wow @luizbon ! Thanks so much for this contribution, and sorry my review was delayed (holidays :sweat_smile: )

I left a few nits which are up to you. I think the only required changes from my perspective are to add a README and [this](https://github.com/open-feature/dotnet-sdk-contrib/pull/119/files#r1438290689).

I've also opened a PR to add you and @laliconfigcat to the owners file [here](https://github.com/prospa-group-oss/dotnet-sdk-contrib/pull/1).

File: src/OpenFeature.Contrib.ConfigCat/ConfigCatProvider.cs
Comment (adams85) [inline src/OpenFeature.Contrib.ConfigCat/ConfigCatProvider.cs:40]: No need for an async state machine here, we can just return the task returned by `ProcessFlag` directly as it's done in the case of the other `Resolve...` methods.

```suggestion
        public override Task<ResolutionDetails<bool>> ResolveBooleanValue(string flagKey, bool defaultValue, EvaluationContext context = null)
        {
            return ProcessFlag(flagKey, context, defaultValue);
        }

```

File: src/OpenFeature.Contrib.ConfigCat/ConfigCatProvider.cs
Comment (adams85) [inline src/OpenFeature.Contrib.ConfigCat/ConfigCatProvider.cs:66]: This implementation looks problematic because this will work only when the feature flag value is a string (for more details, see https://configcat.com/docs/sdk-reference/dotnet/#setting-type-mapping).

However, `Value` can store values of other types than string (see [type definition](https://github.com/open-feature/dotnet-sdk/blob/e8ca1da9ed63df9685ec49a9569e0ec99ba0b3b9/src/OpenFeature/Model/Value.cs)).

To make this work with all the feature flag types which is supported by ConfigCat (boolean, string, int, double), we need something like this:

```suggestion
            var user = context?.BuildUser();
            var result = await Client.GetValueDetailsAsync(flagKey, defaultValue?.AsObject, user);
            var returnValue = result.IsDefaultValue ? defaultValue : new Value(result.Value);
            return new ResolutionDetails<Value>(flagKey, returnValue, ParseErrorType(result.ErrorMessage), errorMessage: result.ErrorMessage, variant: result.VariationId);
```

File: test/OpenFeature.Contrib.ConfigCat.Test/ConfigCatProviderTest.cs
Comment (adams85) [inline test/OpenFeature.Contrib.ConfigCat.Test/ConfigCatProviderTest.cs:69]: Flag overrides will only work with values which can be converted to supported feature flag values (boolean, string, int, double). The ConfigCat SDK doesn't know about the `Value` type, doesn't know how to convert it to a feature flag value, so this will lead to an error.

Using a value of a supported feature flag type solves the issue:

```suggestion
                options => { options.FlagOverrides = BuildFlagOverrides(("example-feature", defaultValue.AsString)); });
```

**Review by adams85 — CHANGES_REQUESTED** [review]
Hi, I'm Adam from ConfigCat, I'm the primary maintainer of the ConfigCat .NET SDK currently. This is why @laliconfigcat asked me to take a look at this PR.

Here are my findings.

File: src/OpenFeature.Contrib.ConfigCat/ConfigCatProvider.cs
Comment (adams85) [inline src/OpenFeature.Contrib.ConfigCat/ConfigCatProvider.cs:27]: One more remark. `ConfigCatClient` implements `IDisposable` and instances of such types should usually be disposed by the consumer as soon as they don't need them any longer.

I'm not sure though how big an issue is this in the case of OpenFeature, what are your guidelines regarding resource management. Probably it's not an actual issue as when we were implementing `ConfigCatClient`, we paid careful attention that client instances can be collected by the GC when no strong references to them are left.

File: test/OpenFeature.Contrib.ConfigCat.Test/ConfigCatProviderTest.cs
Comment (luizbon) [reply test/OpenFeature.Contrib.ConfigCat.Test/ConfigCatProviderTest.cs:98]: Kudos to the ConfigCat team.

File: src/OpenFeature.Contrib.ConfigCat/ConfigCatProvider.cs
Comment (luizbon) [reply src/OpenFeature.Contrib.ConfigCat/ConfigCatProvider.cs:40]: Thanks for picking that up.

File: src/OpenFeature.Contrib.ConfigCat/ConfigCatProvider.cs
Comment (luizbon) [reply src/OpenFeature.Contrib.ConfigCat/ConfigCatProvider.cs:28]: For what it is worth, I have implemented the `IDisposable` interface in the provider class to handle this correctly.

File: src/OpenFeature.Contrib.ConfigCat/ConfigCatProvider.cs
Comment (luizbon) [reply src/OpenFeature.Contrib.ConfigCat/ConfigCatProvider.cs:89]: I left it as is for now since it'll capture the general exception anyway.

File: src/OpenFeature.Contrib.ConfigCat/UserBuilder.cs
Comment (luizbon) [reply src/OpenFeature.Contrib.ConfigCat/UserBuilder.cs:11]: Added a README with that info.

File: src/OpenFeature.Contrib.ConfigCat/UserBuilder.cs
Comment (luizbon) [reply src/OpenFeature.Contrib.ConfigCat/UserBuilder.cs:13]: Added tests for this class

File: src/OpenFeature.Contrib.ConfigCat/UserBuilder.cs
Comment (luizbon) [reply src/OpenFeature.Contrib.ConfigCat/UserBuilder.cs:37]: Added a README with that info.

**Comment by luizbon** [comment]
Hi @toddbaert and @adams85, sorry for the delay, I just came back from holidays.
I think I have addressed all your comments. Can you please review it again?

File: src/OpenFeature.Contrib.ConfigCat/ConfigCatProvider.cs
Comment (adams85) [inline src/OpenFeature.Contrib.ConfigCat/ConfigCatProvider.cs:113]: After taking a closer look at the OpenFeature SDK architecture, specifically at [this PR](https://github.com/open-feature/dotnet-sdk/pull/158), it looks like making the provider class implement `IDisposable` is not the correct approach after all.

Also, a proper `Dispose` implementation would be tricky because of the way ConfigCat SDK manages client instances: 

> You can acquire singleton client instances for your SDK keys using the ConfigCatClient.Get(sdkKey: "<sdkKey>") static factory method. (However, please keep in mind that **subsequent calls to ConfigCatClient.Get() with the same SDK Key return a shared client instance, which was set up by the first call**.)

That is, if you create two provider instances with the same SDK key and then dispose one of them, the underlying client will be disposed in the other one as well. Probably we would need some kind of reference counting to make this work correctly.

So, in light of all this, I say we don't implement `IDisposable` for now. Let's address this issue when the provider shutdown feature lands.

Comment (luizbon) [reply src/OpenFeature.Contrib.ConfigCat/ConfigCatProvider.cs:113]: Yeah, I wasn't aware of the Shutdown implementation. I'll remove the disposable implementation.

**Review by adams85 — APPROVED** [review]

**Comment by beeme1mr** [comment]
@toddbaert could you please review this tomorrow?

@luizbon could you please address the formatting issue when you have a moment?

**Comment by luizbon** [comment]
Thanks for reminding me @beeme1mr, I have applied the format fix.

**Review by toddbaert — APPROVED** [review]
@luizbon Thanks again for your contribution, and your patience with the review process. I am approving, however, **I will request one more change** (I'd do it myself if I could, but I can't modify your branch): please add yourself (and perhaps @laliconfigcat and @adams85 if they are interested) to the [component_owners](https://github.com/open-feature/dotnet-sdk-contrib/blob/main/.github/component_owners.yml) for this component, and it's tests. This will add you to PRs for this component, and really helps share the maintenance load.

@laliconfigcat if you end up making changes to the config-cat SDK to send some kind of error code as you mention [here](https://github.com/open-feature/dotnet-sdk-contrib/pull/119#discussion_r1434556165), please create a PR or issue to improve the handling here if you dont mind! :pray:

**Comment by luizbon** [comment]
Thanks, @toddbaert; just made the change to the component owner file and also found I needed to fix the path and package ID for the project.
It should be good to go now.

**Comment by toddbaert** [comment]
@adams85 @laliconfigcat are either of you interested in being added to the `compoent_owners`? This will add your automatically to any PRs impacting this component. We only require one, so @luizbon 's addition is sufficient (note you will also have to be an org member for this automation to work).

In any case, I will merge this tomorrow EOD unless I hear objections! Thanks @luizbon !

**Comment by laliconfigcat** [comment]
> @adams85 @laliconfigcat are either of you interested in being added to the `compoent_owners`? This will add your automatically to any PRs impacting this component. We only require one, so @luizbon 's addition is sufficient (note you will also have to be an org member for this automation to work).
> 
> In any case, I will merge this tomorrow EOD unless I hear objections! Thanks @luizbon !

Thanks for the suggestion, but I just jumped in for a quick PR review. I think @luizbon is the real component owner here :).

**Pull request merged**
